### ISSUE TITLE
PHP 8.1 no longer returns a resource

### ISSUE THREAD

**Issue by necrogami** [laminas/laminas-ldap#24]
According to the changelog for https://www.php.net/manual/en/function.ldap-connect.php

php 8.1 no longer returns a resource for ldap_connect it now returns a class which is causing src/Ldap.php:839 to fail.

**Comment by michelsp** [comment]
![image](https://user-images.githubusercontent.com/5161978/151042746-9b32f8f1-17d0-4265-9913-493705c9b845.png)
![image](https://user-images.githubusercontent.com/5161978/151042875-c1d0c074-fa5a-47ac-a0c7-ed49b625ca6a.png)

**Comment by Ocramius** [comment]
Can we please have a test case for this?

**Comment by necrogami** [comment]
When updating to PHP 8.1 with PHP's LDAP extension I get is not a resource error when trying to connect to LDAP

**Comment by Ocramius** [comment]
Yes, I'm asking for a test case that runs in CI here, and that prevents this from regressions.

The fact that this BC break wasn't caught, is precisely because an automated test does not cover this.

**Comment by DerpgonCz** [comment]
So, how do we approach this? Or rather, what is the expected outcome?

First I'd decide to either split the code base to "compatible  with 8.1", and "compatible with <8.1", or make a fork in code to handle both cases. I am not sure which is better, as I am not really experienced in this field.

Then, the tests. Do we cover just the connection part? Or that `ldap_connect` returns a resource? Or that both works (in case of the code fork)?


I'd really like to help, although I am not sure what exactly is needed.

**Comment by kevinpapst** [comment]
Are you sure that the test cases do not cover it? 
The PR that integrated PHP 8.1 [said in the description](https://github.com/laminas/laminas-ldap/pull/23#issue-1067569477) `Note that Laminas_Ldap online tests were not available.`. 

Looking [at the test](https://github.com/laminas/laminas-ldap/blob/2.13.x/test/ConnectTest.php#L24) I believe that the connection test would find the problem. But looking at the action ([e.g. here](https://github.com/laminas/laminas-ldap/runs/4390942842?check_suite_focus=true)) I cannot see the env `TESTS_LAMINAS_LDAP_ONLINE_ENABLED`, instead I see 200+ skipped tests. Are the real connection tests probably "only" disabled in Github actions? 

<img width="635" alt="Bildschirmfoto 2022-01-27 um 17 54 02" src="https://user-images.githubusercontent.com/533162/151405857-cfa1c824-b1b9-46a4-8e97-e4b3ef065111.png">

Seems there are two things to work on:
- activating LDAP online tests in CI
- moving the 8 calls to `is_resource()` in LDAP class to a dedicated method that can handle both result types (and probably more necessary refactoring)

**Comment by DerpgonCz** [comment]
On the topic of tests:

Sure enough, during my local testing, a lot of errors for PHPUnit in PHP 8.1, a lot less in PHP 8.0 (not allowing SSL, no seeded data, etc.).

HOWEVER 1, it seems like the tests **were** available, at least [according to Travis](https://app.travis-ci.com/github/laminas/laminas-ldap/builds/242705277).
HOWEVER 2, it tested for `php:master`
HOWEVER 3, `php:master` at the time was not `8.1`, it was `8.0.3`
```
$ php --version
PHP 8.0.3-dev (cli) (built: Feb 11 2021 11:15:07) ( ZTS )
```
HOWEVER 4, the current up-to-date `2.13.x` branch [does not go through Travis](https://github.com/laminas/laminas-ldap/commit/96ed993f3958e85791e27c9bfa175b53c90a4866#diff-6ac3f79fc25d95cd1e3d51da53a4b21b939437392578a35ae8cd6d5366ca5485) due to [change in CI workflow](https://github.com/laminas/laminas-ldap/commit/10e994176552028bbe64dc1a6be3e24edb17e301) which [was part of said PHP 8.1 PR](https://github.com/laminas/laminas-ldap/commit/4313a5f6a12b8a78ed1c2b79b78a636898ffebaa), only `2.12.0` [went through Travis](https://app.travis-ci.com/github/laminas/laminas-ldap/builds/212863034), and tested for LDAP, but it does not run for PHP 8.1.

I think that refactoring CI overlooked LDAP tests.

So, basically two things needs to be done:

- [ ] Run Online LDAP tests in the current CI configuration
- [ ] Fix the error

As far as fixing the error goes, I am suspecting that checking if `is_resource` could just be changed to check for `!== false` (as stated [here](https://php.watch/versions/8.1/LDAP-resource) for example), as other `ldap_*` functions changed their signatures from `resource`s to `LDAP\Connection`, `LDAP\Result` and `LDAP\ResultEntry`. Those were not directly used, only passed along to other `ldap_*` functions.

**Comment by kevinpapst** [comment]
@Ocramius having LDAP is 8.1 is rather crucial. What can we do, how can we help? We already provided some feedback, but I can support with testing as I have automated tests that really use LDAP connections in combination with this package, but I don't feel confident enough to fix the issue here. 

I wonder what @Koen1999 who worked on the PHP 8.1 PR and @froschdesign who asked for the support initially think about it.

**Comment by Ocramius** [comment]
@DerpgonCz Travis-CI is dead-dead-dead. 

The last CI build on `2.12.x` was on https://github.com/laminas/laminas-ldap/runs/4407780273?check_suite_focus=true, and all tests passed: https://github.com/laminas/laminas-ldap/runs/4407780273?check_suite_focus=true#step:3:346, so yep, we need to add a running LDAP server to CI :) 

So here again: provide a failing test case (**FIRST**), then a fix, if possible, and it will gladly be reviewed/merged.

@kevinpapst if it's crucial, put dev time in it and help as described in this thread: fork the repo, get CI running with a real server, send a patch. I know as much as you do ;-)

**Comment by kevinpapst** [comment]
I know you are a busy person, but you are not even reading our comments. The tests were all skipped, so it is extremely likely that they already exist. This is from your link:

<img width="745" alt="Bildschirmfoto 2022-02-10 um 12 15 47" src="https://user-images.githubusercontent.com/533162/153396307-add34250-4d6e-47c8-9519-c795b5411f09.png">

**Comment by froschdesign** [comment]
@kevinpapst 
> I cannot see the env `TESTS_LAMINAS_LDAP_ONLINE_ENABLED`

Check the configuration of PHPUnit:

https://github.com/laminas/laminas-ldap/blob/5e0debb0c2e38f5c197eacf7a8d99b79f74e1110/phpunit.xml.dist#L46

**Comment by Ocramius** [comment]
> but you are not even reading our comments.

What I said: 

 > so yep, we need to add a running LDAP server to CI :)

**Comment by kevinpapst** [comment]
@Ocramius I do not understand the Laminas CI actions/docker images. So could you please join and help me with #25 ?
The tests now fail in every PHP version, but the higher versions have way more failing tests.

**Comment by Koen1999** [comment]
> I wonder what @Koen1999 who worked on the PHP 8.1 PR and @froschdesign who asked for the support initially think about it.

Honestly I don't think much about it. When I made that PR, I was trying to add PHP 8.1 support for another Laminas package that somehow required laminas-ldap even though it was optional. Since the test suite passed for the PR, I assumed this would work. I don't know much about LDAP, but if this package is blocking you I suggest you stick with PHP 8.0 for now and try to work on a fix for the issue.

**Comment by kevinpapst** [comment]
After spending 2 hours now to get the test cases up and running, I can share:
- they run and they show that the issue is covered in 8.1
- they still have an issue though in lower versions, likely because there is an issue with importing iptable rules which I cannot fix

So if someone wants to join and help fixing 8.1, please post your ideas or fork my branch.

**Comment by bchen-us** [comment]
> On the topic of tests:
> 
> Sure enough, during my local testing, a lot of errors for PHPUnit in PHP 8.1, a lot less in PHP 8.0 (not allowing SSL, no seeded data, etc.).
> 
> HOWEVER 1, it seems like the tests **were** available, at least [according to Travis](https://app.travis-ci.com/github/laminas/laminas-ldap/builds/242705277). HOWEVER 2, it tested for `php:master` HOWEVER 3, `php:master` at the time was not `8.1`, it was `8.0.3`
> 
> ```
> $ php --version
> PHP 8.0.3-dev (cli) (built: Feb 11 2021 11:15:07) ( ZTS )
> ```
> 
> HOWEVER 4, the current up-to-date `2.13.x` branch [does not go through Travis](https://github.com/laminas/laminas-ldap/commit/96ed993f3958e85791e27c9bfa175b53c90a4866#diff-6ac3f79fc25d95cd1e3d51da53a4b21b939437392578a35ae8cd6d5366ca5485) due to [change in CI workflow](https://github.com/laminas/laminas-ldap/commit/10e994176552028bbe64dc1a6be3e24edb17e301) which [was part of said PHP 8.1 PR](https://github.com/laminas/laminas-ldap/commit/4313a5f6a12b8a78ed1c2b79b78a636898ffebaa), only `2.12.0` [went through Travis](https://app.travis-ci.com/github/laminas/laminas-ldap/builds/212863034), and tested for LDAP, but it does not run for PHP 8.1.
> 
> I think that refactoring CI overlooked LDAP tests.
> 
> So, basically two things needs to be done:
> 
>     * [ ]  Run Online LDAP tests in the current CI configuration
> 
>     * [ ]  Fix the error
> 
> 
> As far as fixing the error goes, I am suspecting that checking if `is_resource` could just be changed to check for `!== false` (as stated [here](https://php.watch/versions/8.1/LDAP-resource) for example), as other `ldap_*` functions changed their signatures from `resource`s to `LDAP\Connection`, `LDAP\Result` and `LDAP\ResultEntry`. Those were not directly used, only passed along to other `ldap_*` functions.

I find that adding isset($this->$resource) works as well. For most of the code I use 'isset($this->$resource)' and check for '!== false'. In some cases, I check for isset($this->$resource) after since certain functions will instantiate the LDAP objects and will want to check if it was set after calling certain functions.

**Comment by Ocramius** [comment]
Please no hacky workarounds: help us get the package up to date in #25!

**Comment by kevinpapst** [comment]
#25 closed in favor of #29 and now the code needs to be adjusted => in #27

**Comment by Ocramius** [comment]
Releasing `2.13.0` - please give it a shot and see if that fixes the problem, y'all :)

**Comment by arueckauer** [comment]
It looks like this is fixed and and can be closed.
